### PR TITLE
Feature/add great expectations operator gex ops-227

### PR DESCRIPTION
**Pull request by AnthonyByansi**
**_WHAT DOES THIS PR DO?_**
This PR introduces a new checkpoint functionality

**_WHAT ISSUES ARE RELATED TO THIS PR?_**

- Jira cards
 OPS 227
 
**_ARE THERE ANY RELATED PRs?_**
Related PR: #3262

## Summary by CodeRabbit

- **New Features**
  - Introduced the `GreatExpectationsOperator` for integrating Great Expectations validation into Airflow workflows.
  - Added a new task for data validation within the workflow, enhancing data quality assurance.
  - Launched the `AirQoGx` class to streamline data validation management within ETL processes.

- **Enhancements**
  - Updated functions to use dynamic placeholders for data source and dataset configurations, increasing flexibility.

- **Dependency Updates**
  - Updated several dependencies, including `great_expectations`, `airflow-provider-great-expectations`, `pytest`, `sqlalchemy`, `pandas-gbq`, and `google-cloud-bigquery`.

### CODE REVIEW

File: gx/operators/great_expectations.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
+import os
+from datetime import datetime
+from pathlib import Path
```
[the remark this answers is not in the document]
Comment (AnthonyByansi) [reply gx/operators/great_expectations.py]: resolved at: https://github.com/airqo-platform/AirQo-api/pull/3263/commits/64dceb6c7282e0200e2490c1cbf55bf3207c60ac

**Review by Baalmart — APPROVED** [review]
Thanks @AnthonyByansi , all looks good to me.
Please coordinate others in the background to review your PR.

File: gx/operators/great_expectations.py [hunk 2]
```diff
@@ -0,0 +1,618 @@
...
+        self.checkpoint_kwargs: Optional[Dict[str, Any]] = checkpoint_kwargs
+        self.fail_task_on_validation_failure: Optional[bool] = (
+            fail_task_on_validation_failure
+        )
+        self.validation_failure_callback: Optional[
+            Callable[[CheckpointResult], None]
+        ] = validation_failure_callback
+        self.return_json_dict: bool = return_json_dict
+        self.use_open_lineage = use_open_lineage
+        self.is_dataframe = True if self.dataframe_to_validate is not None else False
```
[the remark this answers is not in the document]
Comment (AnthonyByansi) [reply gx/operators/great_expectations.py]: Resolved through: https://github.com/airqo-platform/AirQo-api/pull/3263/commits/e9c29645d8226f5f3b35f1721e147f4b79bfc9d4

File: gx/operators/great_expectations.py [hunk 3]
```diff
@@ -0,0 +1,618 @@
...
+            context["ti"].xcom_push(key="data_docs_url", value=data_docs_site)
+        except KeyError:
+            self.log.debug("Could not push data_docs_url to XCom.")
+        self.log.info("GE Checkpoint Run Result:\n%s", result)
+        self.handle_result(result)
+
+        if self.return_json_dict:
+            return result.to_json_dict()
+
+        return result
```
[the remark this answers is not in the document]
Comment (AnthonyByansi) [reply gx/operators/great_expectations.py]: Resolved through: https://github.com/airqo-platform/AirQo-api/pull/3263/commits/3bdb1cb03452a6cd80e61821043ceeef21125526

File: gx/operators/great_expectations.py [hunk 4]
```diff
@@ -0,0 +1,618 @@
...
+                raise ValueError(
+                    "Specify the name of the database in the schema parameter of the Postgres connection. See: https://airflow.apache.org/docs/apache-airflow-providers-postgres/stable/connections/postgres.html"  # noqa
+                )
+
+        elif conn_type == "gcpbigquery":
+            uri_string = f"{self.conn.host}{self.schema}"
+        elif conn_type == "sqlite":
+            uri_string = f"sqlite:///{self.conn.host}"
+        else:
+            raise ValueError(f"Conn type: {conn_type} is not supported.")
```
[the remark this answers is not in the document]
Comment (AnthonyByansi) [reply gx/operators/great_expectations.py]: Resolved

**Review by NicholasTurner23 — CHANGES_REQUESTED** [review]
Hi @AnthonyByansi 

This code has had a couple of iterations. Could you look at the current version in staging and may be make changes to that?

In fact the gx folder that was on the same level as the src folder was deleted. I would suggest closing this branch and open another.

**Pull request closed without merging**